contracts/modules/cover/Cover.sol

```diff
@@ -0,0 +1,570 @@
+
+import "@openzeppelin/contracts-v4/token/ERC721/ERC721.sol";
+import "@openzeppelin/contracts-v4/token/ERC20/IERC20.sol";
+import "../../interfaces/ICover.sol";
+import "../../interfaces/IStakingPool.sol";
+import "../../interfaces/IPool.sol";
+import "../../abstract/MasterAwareV2.sol";
+import "../../interfaces/IMemberRoles.sol";
+import "hardhat/console.sol";
+
+
+contract Cover is ICover, ERC721, MasterAwareV2 {
+
+  Cover[] public override covers;
+  mapping(uint => StakingPool[]) stakingPoolsForCover;
+
+  Product[] public override products;
+  mapping(uint => uint) capacityFactors;
+
+  ProductType[] public override productTypes;
+
+  mapping(uint => uint) initialPrices;
+
+  mapping(uint => uint96) public override activeCoverAmountInNXM;
+
+  /*
+   (productId, poolAddress) => lastPrice
+   Last base prices at which a cover was sold by a pool for a particular product.
+  */
+  mapping(uint => mapping(address => uint)) lastPrices;
+
+  /*
+   (productId, poolAddress) => lastPriceUpdate
+   Last base price update time.
+  */
+  mapping(uint => mapping(address => uint)) lastPriceUpdate;
+
+
+  /* === CONSTANTS ==== */
+
+  uint public REWARD_BPS = 5000;
+  uint public constant PERCENTAGE_CHANGE_PER_DAY_BPS = 100;
+  uint public constant BASIS_PRECISION = 10000;
+  uint public constant STAKE_SPEED_UNIT = 100000e18;
+  uint public constant PRICE_CURVE_EXPONENT = 7;
+  uint public constant MAX_PRICE_PERCENTAGE = 1e20;
+
+  /* === MODIFIERS ==== */
+
+  modifier onlyAdvisoryBoard {
+    uint abRole = uint(IMemberRoles.Role.AdvisoryBoard);
+    require(
+      memberRoles().checkRole(msg.sender, abRole),
+      "Cover: Caller is not an advisory board member"
+    );
+    _;
+  }
+
+  constructor(string memory name_, string memory symbol_) ERC721(name_, symbol_) {
+  }
+
+  /* === MUTATIVE FUNCTIONS ==== */
+
+  function buyCover(
+    address owner,
+    uint24 productId,
+    uint8 payoutAsset,
+    uint96 amount,
+    uint32 period,
+    uint maxPrice,
+    StakingPool[] memory stakingPools
+  ) external payable override returns (uint /*coverId*/) {
+    require(initialPrices[productId] != 0, "Cover: product not initialized");
+
+    (uint coverId, uint priceInAsset) = _createCover(owner, productId, payoutAsset, amount, period, stakingPools);
+    require(priceInAsset <= maxPrice, "Cover: Price exceeds maxPrice");
+    retrievePayment(priceInAsset, payoutAsset);
+    return coverId;
+  }
+
+  function _createCover(
+    address owner,
+    uint24 productId,
+    uint8 payoutAsset,
+    uint96 amount,
+    uint32 period,
+    StakingPool[] memory stakingPools
+  ) internal returns (uint coverId, uint premiumInAsset) {
+
+    // convert to NXM amount
+    uint amountLeftToCoverInNXM = uint(amount) * 1e18 / pool().getTokenPrice(pool().assets(payoutAsset));
```

Comment:
`getTokenPrice` and `pool.assets()` are called multiple times, will need to store them in variables at the optimization stage.

Reply:
`amountLeftToCoverInNXM` -> `coverAmountinNXM`